src/equihash.rs

```diff
@@ -0,0 +1,471 @@
+                    & (
+                    // Apply bit_len_mask across byte boundaries
+                    (bit_len_mask >> (8 * (out_width - x - 1))) & 0xFF
+                )) as u8;
+            }
+            j += out_width;
+        }
+    }
+
+    vout
+}
+
+fn indices_from_minimal(minimal: &[u8], c_bit_len: usize) -> Vec<u32> {
+    assert!(((c_bit_len + 1) + 7) / 8 <= size_of::<u32>());
+    let len_indices = 8 * size_of::<u32>() * minimal.len() / (c_bit_len + 1);
+    let byte_pad = size_of::<u32>() - ((c_bit_len + 1) + 7) / 8;
+
+    let mut csr = Cursor::new(expand_array(minimal, c_bit_len + 1, byte_pad));
+    let mut ret = Vec::with_capacity(len_indices);
+
+    // Big-endian so that lexicographic array comparison is equivalent to integer
+    // comparison
+    while let Ok(i) = csr.read_u32::<BigEndian>() {
+        ret.push(i);
+    }
+
+    ret
+}
+
+fn has_collision(a: &Node, b: &Node, len: usize) -> bool {
+    a.hash
+        .iter()
+        .zip(b.hash.iter())
+        .take(len)
+        .all(|(a, b)| a == b)
+}
+
+fn distinct_indices(a: &Node, b: &Node) -> bool {
+    for i in &(a.indices) {
+        for j in &(b.indices) {
+            if i == j {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+pub fn is_valid_solution_iterative(
+    n: u32,
+    k: u32,
+    input: &[u8],
+    nonce: &[u8],
+    indices: &[u32],
+) -> bool {
+    let p = Params { n: n, k: k };
+
+    let mut state = initialise_state(p.n, p.k, p.hash_output());
+    state.update(input);
+    state.update(nonce);
+
+    let mut rows = Vec::new();
+    for i in indices {
+        let hash = generate_hash(&state, i / p.indices_per_hash_output());
+        let start = ((i % p.indices_per_hash_output()) * p.n / 8) as usize;
+        let end = start + (p.n as usize) / 8;
+        rows.push(Node {
+            hash: expand_array(&hash.as_bytes()[start..end], p.collision_bit_length(), 0),
+            indices: vec![*i],
+        });
+    }
+
+    let mut hash_len = p.hash_length();
+    while rows.len() > 1 {
+        let mut cur_rows = Vec::new();
+        for pair in rows.chunks(2) {
+            let a = &pair[0];
+            let b = &pair[1];
+            if !has_collision(a, b, p.collision_byte_length()) {
+                // error!("Invalid solution: invalid collision length between StepRows");
+                return false;
+            }
+            if b.indices_before(a) {
+                // error!("Invalid solution: Index tree incorrectly ordered");
+                return false;
+            }
+            if !distinct_indices(a, b) {
+                // error!("Invalid solution: duplicate indices");
+                return false;
+            }
+            cur_rows.push(Node::from_children_ref(a, b, p.collision_byte_length()));
+        }
+        rows = cur_rows;
+        hash_len -= p.collision_byte_length();
+    }
+
+    assert!(rows.len() == 1);
+    return rows[0].is_zero(hash_len);
```

Comment:
Why do we use ```hash_len``` here rather than ```p.collision_byte_length()```?

Reply:
The final round collides two chunks instead of one, and since we've been subtracting from `hash_len` along the way, it's the correct length, so it's fine to say "just check the entire remaining hash is zero".